cni/pkg/ebpf/ambient_redirect_kern.c

```diff
@@ -0,0 +1,319 @@
+// ip netns exec ${NS} tc filter add dev eth0 ingress bpf da obj ambient_redirect.o sec ztunnelingress
+
+#include <linux/bpf.h>
+#include <linux/pkt_cls.h>
+#include <netinet/in.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <linux/if_packet.h>
+#include <linux/if_ether.h>
+#include <linux/tcp.h>
+#include <stdint.h>
+#include <iproute2/bpf_elf.h>
+#include <bpf/bpf_helpers.h>
+#include <string.h>
+#include "common.h"
+
+
+#ifndef __section
+# define __section(NAME)                  \
+   __attribute__((section(NAME), used))
+#endif
+
+#ifndef __inline
+# define __inline                         \
+   inline __attribute__((always_inline))
+#endif
+
+
+#ifndef BPF_FUNC
+# define BPF_FUNC(NAME, ...)              \
+   (*NAME)(__VA_ARGS__) = (void *)BPF_FUNC_##NAME
+#endif
+
+static void *BPF_FUNC(map_lookup_elem, void *map, const void *key);
+static void BPF_FUNC(trace_printk, const char *fmt, int fmt_size, ...);
+
+#ifndef printk
+#define printk(fmt, ...)                                      \
+    ({                                                         \
+        char ____fmt[] = fmt;                                  \
+        trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
+    })
+#endif
+
+#define dbg(fmt, ...)                                      \
+    ({                                                         \
+        char ____fmt[] = fmt;                                  \
+        trace_printk(____fmt, sizeof(____fmt), ##__VA_ARGS__); \
+    })
+
+
+
+struct bpf_elf_map ztunnel_info __section("maps") = {
+    .type           = BPF_MAP_TYPE_ARRAY,
+    .size_key       = sizeof(uint32_t),
+    .size_value     = sizeof(struct ztunnel_info),
+    .pinning        = PIN_GLOBAL_NS,
+    .max_elem       = 1,
+};
+
+struct bpf_elf_map app_info __section("maps") = {
+    .type           = BPF_MAP_TYPE_HASH,
+    .size_key       = sizeof(uint32_t),
+    .size_value     = sizeof(struct app_info),
+    .pinning        = PIN_GLOBAL_NS,
+    .max_elem       = APP_INFO_MAP_SIZE,
+};
+
+
+static __inline struct ztunnel_info * get_ztunnel_info()
+{
+    uint32_t key = 0;
+    return map_lookup_elem(&ztunnel_info, &key);
+}
+
+static __inline struct app_info * get_app_info_from_ipv4(uint32_t ipv4)
+{
+    return map_lookup_elem(&app_info, &ipv4);
+}
+
+__section("ztunnelingress")
+int ztunnel_ingress(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    struct eth_hdr *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct bpf_sock_tuple *tuple;
+    size_t tuple_len;
+    struct bpf_sock *sk;
+    int skip_mark = 0;
+    __u32 outbound_cb = OUTBOUND_CB;
+    __u32 inbound_cb = INBOUND_CB;
+
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    if (eth->h_proto != htons(ETH_P_IP)) {
+        return TC_ACT_OK;
+    }
+    struct iphdr *iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end) {
```

Comment:
`iph + 1 > data_end`

Reply:
`iph + 1 `   might cause compile warning in some cases, so let's keep current approach